demo/api/src/importer/pipes/parsers/data-transformer.pipe.ts

```diff
@@ -0,0 +1,209 @@
+import { Transform, TransformCallback } from "stream";
+
+import { ImporterPipe, PipeData, ValidationError } from "../importer-pipe.type";
+
+type ValidateAndTransformValueProperties = {
+    value: string;
+    propertyMetadata: TargetEntityProperty;
+    field: ImportFieldMetadata;
+};
+
+type DataTransformOptions = {
+    fields: ImportFieldMetadata[];
+    targetEntityProperties: TargetEntityProperties;
+    entityDateFormatString?: string;
+    fieldTransformers?: FieldTransformerData[];
+};
+type PipeDataAndErrors = {
+    data: PipeData;
+    errors: ValidationError[];
+};
+
+type PossibleValueTypes = string | number | boolean | Date | undefined | null | object;
+
+export class DataTransformerPipe implements ImporterPipe {
+    private readonly options: DataTransformOptions;
+
+    constructor(options: DataTransformOptions, private readonly entity: ImporterEntityClass) {
+        this.options = options;
+    }
+
+    getPipe(runLogger: LoggerService) {
+        return new DataTransformer(this.options, runLogger, this.entity);
+    }
+}
+
+type ParserPipeData = Record<string, string>;
+
+export class DataTransformer extends Transform {
+    constructor(
+        private readonly options: DataTransformOptions,
+        private readonly logger: LoggerService,
+        private readonly entity: ImporterEntityClass,
+    ) {
+        super({ writableObjectMode: true, objectMode: true });
+        this.logger = logger;
+        this.options = options;
+    }
+
+    async _transform(inputData: ParserPipeData, encoding: BufferEncoding, callback: TransformCallback) {
+        try {
+            const transformedDataAndErrors = this.transformData(inputData);
+            if (transformedDataAndErrors.errors.length > 0) {
+                transformedDataAndErrors.errors.forEach((error) => {
+                    this.logger.error(`Transformation Error: ${JSON.stringify(error)}`);
+                });
+                this.push(null);
+                return callback(new Error("Too many transformation errors"));
+            }
+            const { data, errors } = await this.convertToInstanceAndValidate(transformedDataAndErrors.data);
+            if (errors.length > 0) {
+                errors.forEach((error) => {
+                    this.logger.error(`Validation Errors: ${JSON.stringify(error)}`);
+                });
+                this.push(null);
+                return callback(new Error("Too many validation errors"));
+            }
+            this.push(data);
+            // eslint-disable-next-line  @typescript-eslint/no-explicit-any
+        } catch (err: any) {
+            this.logger.error(`Error transforming Data: ${err}`);
+            this.emit("error", err);
+            return callback(err);
+        }
+
+        callback();
+    }
+
+    async convertToInstanceAndValidate(data: PipeData): Promise<PipeDataAndErrors> {
+        const instance = plainToInstance(this.entity, data) as Record<string, unknown>;
```

Comment:
The MikroORM maintainer notes that using class-transformer to create an entity is ["a very bad idea"](https://github.com/mikro-orm/mikro-orm/discussions/6353#discussioncomment-11927293). We should discuss this if we should add an additional DTO layer instead.

Reply:
Meaning: The class-transformer creates an instance but can't initialize Refs and Collections so the typing isn't correct anymore. Do I understand that correctly?

Having a DTO layer means the typing is correct but we again have to define the "Product" twice (like we did in the RawProduct that we removed)

Reply:
Also, now the fields are automatically validated against the DB field type. We'd lose that when we define a DTO. Then we can only rely on class-validator decorators. Persisting in the DB would still fail but only in the PersistPipe